source/4_tools/regina/src/regina.cpp

```diff
@@ -1,54 +1,865 @@
+        create_clear_state_command_desc.depth = 1.0f;
+        create_clear_state_command_desc.stencil = 0x00;
+
+        StateController<renderer::ClearBits> clear_flags;
+        clear_flags.add_state(renderer::ClearBits::CLEAR_COLOR_BUFFER);
+        clear_flags.add_state(renderer::ClearBits::CLEAR_DEPTH_BUFFER);
+        clear_flags.add_state(renderer::ClearBits::CLEAR_STENCIL_BUFFER);
+
+        create_clear_state_command_desc.flags = clear_flags;
+
+        g_regina_ctx.clear_state = renderer::create_clear_state(rsl::move(create_clear_state_command_desc));
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+    bool build_shader_and_input_layout()
+    {
+        // Shader
+        renderer::commands::CompileShaderCommandDesc vs_compile_command_desc = create_compile_shader_parameters(rsl::small_stack_string("standardVS"), renderer::ShaderType::VERTEX, "Shaders\\color.hlsl");
+        renderer::commands::CompileShaderCommandDesc ps_compile_command_desc = create_compile_shader_parameters(rsl::small_stack_string("opaquePS"), renderer::ShaderType::PIXEL, "Shaders\\color.hlsl");
+
+        renderer::commands::LinkShaderCommandDesc link_shader_command_desc;
+        link_shader_command_desc.vertex_shader = renderer::compile_shader(rsl::move(vs_compile_command_desc));
+        link_shader_command_desc.pixel_shader = renderer::compile_shader(rsl::move(ps_compile_command_desc));
+        link_shader_command_desc.constants =
+        {
+            renderer::commands::ConstantLayoutDescription {renderer::commands::ConstantType::CBUFFER, "ObjectConstants", 0},
+            renderer::commands::ConstantLayoutDescription {renderer::commands::ConstantType::CBUFFER, "PassConstants", 1}
+        };
+        g_regina_ctx.shader_program = renderer::link_shader(rsl::move(link_shader_command_desc));
+
+        // Input Layout
+        renderer::commands::CreateInputLayoutCommandDesc input_layout_command_desc;
+        input_layout_command_desc.input_layout =
+        {
+            renderer::commands::InputLayoutDescription { "POSITION", 0, renderer::VertexBufferFormat::FLOAT3, 0, 0, renderer::InputLayoutClassification::PER_VERTEX_DATA, 0 },
+            renderer::commands::InputLayoutDescription { "COLOR", 0, renderer::VertexBufferFormat::FLOAT4, 0, 12, renderer::InputLayoutClassification::PER_VERTEX_DATA, 0 }
+        };
+        g_regina_ctx.input_layout = renderer::create_input_layout(rsl::move(input_layout_command_desc));
+
+        return true;
+    }
+
+    //-------------------------------------------------------------------------
+    bool build_cube_geometry()
+    {
+        auto box = mesh_factory::create_box<u16>(1.5f, 0.5f, 1.5f, 0);
+
+        auto total_vertex_count = box.vertices().size();
+        auto total_index_count = box.indices().size();
+
+        rsl::vector<renderer::VertexPosCol> box_vertices((rsl::Capacity)total_vertex_count);
```

Comment:
that's a weird way of using Capacity.. you can/should use it like this rsl::Capacity(total_vertex_count)
also, why are you not just doing: const rsl::vector<u16>& box_indices = box.indices(); ?

Reply:
I cannot use rsl::Capacity as rsl::Capacity(num) cuz then the code does not compile.
box_indices and box.indices() have a different type ( this is the issue I addressed previously with the MeshFactory ) the Vertex type that is used within the MeshFactory should align with the vertex type we use for the renderer only now it isn't